tudatBundle/pagmo2/include/pagmo/algorithms/giACOmo.hpp

```diff
@@ -0,0 +1,600 @@
+        // We start by checking that the problem is suitable for this
+        // particular algorithm.
+
+        if (!NP) {
+            pagmo_throw(std::invalid_argument, get_name() + " cannot work on an empty population");
+        }
+        if (prob.is_stochastic()) {
+            pagmo_throw(std::invalid_argument,
+                        "The problem appears to be stochastic " + get_name() + " cannot deal with it");
+        }
+        if (prob.get_nc() != 0u) {
+            pagmo_throw(std::invalid_argument, "Non linear constraints detected in " + prob.get_name() + " instance. "
+                                                   + get_name() + " cannot deal with them.");
+        }
+        if (m_gen == 0u){
+            return pop;
+        }
+
+
+
+        // ---------------------------------------------------------------------------------------------------------
+
+        // No throws, all valid: we clear the logs
+        m_log.clear();
+
+        // Declarations
+
+
+        // Main ACO loop over generations:
+        for (decltype(m_gen) gen = 1u; gen <= m_gen; gen++) {
+            // 0 - Logs and prints (verbosity modes > 1: a line is added every m_verbosity generations)
+            if (m_verbosity > 0u) {
+                // Every m_verbosity generations print a log line
+                if (gen % m_verbosity == 1u || m_verbosity == 1u) {
+                    // We compute the ideal point
+                    vector_double ideal_point = ideal(pop.get_f());
+                    // Every 50 lines print the column names
+                    if (count % 50u == 1u) {
+                        print("\n", std::setw(7), "Gen:", std::setw(15), "Fevals:");
+                        for (decltype(ideal_point.size()) i = 0u; i < ideal_point.size(); ++i) {
+                            if (i >= 5u) {
+                                print(std::setw(15), "... :");
+                                break;
+                            }
+                            print(std::setw(15), "ideal" + std::to_string(i + 1u) + ":");
+                        }
+                        print('\n');
+                    }
+                    print(std::setw(7), gen, std::setw(15), prob.get_fevals() - fevals0);
+                    for (decltype(ideal_point.size()) i = 0u; i < ideal_point.size(); ++i) {
+                        if (i >= 5u) {
+                            break;
+                        }
+                        print(std::setw(15), ideal_point[i]);
+                    }
+                    print('\n');
+                    ++count;
+                    // Logs
+                    m_log.emplace_back(gen, prob.get_fevals() - fevals0, ideal_point);
+                }
+            }
+
+            // At each generation we make a copy of the population into popnew
+            population popnew(pop);
+
+            //In the case the algorithm is multi-objective, a decomposition strategy is applied:
```

Comment:
So, let me understand, if the problem has, say, 2 objectives f1, f2, this algo runs the single objective evolve on some decomposed fitness (for example f1+f2)?

Reply:
Yes, but I am not sure yet whether I will use a single fitness value directly (e.g. f1+f2) or if I will try to implement first a parallelization strategy in order to first solve some sub-problems as separate single objective problems, for then using a strategy in order to proceed (e.g. utopia-nadir-balance decomposition). The second method seems to be more promising maybe. What do you think about it? Do you have any suggestions for the decomposition strategy?

Reply:
In Pagmo decomposition is implemented both as utilities: https://esa.github.io/pagmo2/docs/cpp/utils/multi_objective.html and as a meta problem.

In any case, I would leave out the multiobjective part for the tiem being. Just develop a single objective unconsrained solver. Then We can discuss on how to add the various extensions to it.

Reply:
Okay, sounds like a plan.